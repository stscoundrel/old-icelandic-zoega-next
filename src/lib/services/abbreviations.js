import { findAbbreviations } from 'old-icelandic-zoega-abbreviations'
import { abbreviate } from 'abbreviatrix'

export const getAbbreviations = ({ definitions }) => {
  const combinedAbbreviations = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = findAbbreviations(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (content, abbreviations) => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
