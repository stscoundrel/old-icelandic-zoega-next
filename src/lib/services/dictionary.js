const { getDictionary } = require('old-icelandic-zoega')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let dictionaryCache = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (dictionaryCache) return dictionaryCache

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  dictionaryCache = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
