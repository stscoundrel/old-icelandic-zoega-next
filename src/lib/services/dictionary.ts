import { getDictionary, getNoMarkupDictionary } from 'old-icelandic-zoega'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import type { DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

// Light weight alternative when definitions are not needed.
// Next.js caches all props as JSON, so no need to store megabytes of definitions when not used.
export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let dictionaryCache: DictionaryEntry[] | null = null
let cachedInitialPages: string[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word)

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (dictionaryCache) return dictionaryCache

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  dictionaryCache = formattedWords

  return formattedWords
}

export const getAllWordsWithoutMarkup = (): DictionaryEntry[] => {
  const words = getNoMarkupDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  return formattedWords
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => ({ word: entry.word, slug: entry.slug }))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWordsWithoutMarkup()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

/**
 * Initial word pages to build are basically 6000
 * headword pages based on modulus. Larger number
 * can not be deployed in one go.
 */
export const getInitialWordsToBuild = (): string[] => {
  if (cachedInitialPages) return cachedInitialPages

  const allWords = getAllWords()

  const result: string[] = []
  for (let i = 0; i < allWords.length; i += 5) {
    result.push(allWords[i].slug);
  }

  cachedInitialPages = result
  return cachedInitialPages
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
