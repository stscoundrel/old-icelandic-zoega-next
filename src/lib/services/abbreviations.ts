import { findAbbreviations } from 'old-icelandic-zoega-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from './dictionary'

export interface Abbreviation{
  abbreviation: string,
  explanation: string
}

export const getAbbreviations = ({ definitions }: DictionaryEntry): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = findAbbreviations(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations:
  Abbreviation[],
): string => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
