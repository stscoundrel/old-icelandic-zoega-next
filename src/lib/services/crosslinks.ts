import { getCrosslinks as getAllCrossLinks, Crosslink } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from './dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    // Only return foreign links, not ones pointing to self.
    return crosslinks[entry.slug].filter((link) => link.source !== 'old-icelandic')
  }

  return []
}

export default {
  getCrossLinks,
}
