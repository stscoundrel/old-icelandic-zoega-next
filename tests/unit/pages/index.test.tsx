import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import { matchesSchema } from 'jafningjar'
import Index, { getStaticProps } from 'pages/index'
import { getAllWords, getAlphabet } from 'lib/services/dictionary'

describe('Index page', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Index words={getAllWords().slice(0, 100)} letters={getAlphabet()} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <Index words={getAllWords().slice(0, 100)} letters={getAlphabet()} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Index page: data fetching', () => {
  test('getStaticProps works', async () => {
    const expected = {
      props: {
        letters: [],
        words: [],
      },
    }

    const result = await getStaticProps()

    expect(matchesSchema(result, expected)).toBeTruthy()
  })
})
