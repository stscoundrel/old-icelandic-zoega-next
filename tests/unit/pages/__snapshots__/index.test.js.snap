// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page Matches snapshot 1`] = `
Array [
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "margin": 0,
              "maxWidth": "100%",
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "block",
                "maxWidth": "100%",
              }
            }
          >
            <img
              alt=""
              aria-hidden={true}
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMzAiIGhlaWdodD0iMzAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
              style={
                Object {
                  "border": "none",
                  "display": "block",
                  "margin": 0,
                  "maxWidth": "100%",
                  "padding": 0,
                }
              }
            />
          </div>
          <noscript>
            <img
              alt="To home"
              decoding="async"
              onClick={[Function]}
              onMouseEnter={[Function]}
              src="/_next/image?url=%2Ffavicon-48x48.png&w=64&q=75"
              srcSet="/_next/image?url=%2Ffavicon-48x48.png&w=32&q=75 1x, /_next/image?url=%2Ffavicon-48x48.png&w=64&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </noscript>
          <img
            alt="To home"
            decoding="async"
            onClick={[Function]}
            onMouseEnter={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </div>
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              á
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              é
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              í
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              ó
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              ú
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              ý
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/th"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              þ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              æ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe3"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              œ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              ö
            </a>
          </li>
        </ul>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <input
            aria-label="Search"
            className="input"
            name="search"
            onChange={[Function]}
            type="search"
            value=""
          />
          <button
            className="button button--secondary button--no-margs"
            name="submit"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Old Icelandic Dictionary
        </a>
        <script
          dangerouslySetInnerHTML={
            Object {
              "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Old Icelandic Dictionary\\",\\"item\\":\\"undefined/\\"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <section
      className="section"
    >
      <h1
        className="h2"
      >
        A Concise Dictionary of Old Icelandic
      </h1>
      <p>
        Online version of the \`"A Concise Dictionary of Old Icelandic\`" dictionary by Geir Zoëga, originally published in 1910
      </p>
      <p>
        Zoëgas attempt was to made easier-to-approach version of the more full Cleasby - Vigfusson dictionary, specifically for beginners and those interested in Old Icelandic prose writing.
      </p>
      <a
        className="button"
        href="/search"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Search the dictionary
      </a>
    </section>
    <section
      className="section"
    >
      <h2
        className="h3"
      >
        What is Old Icelandic?
      </h2>
      <p>
        Old Icelandic was a western dialect of Old Norse. Old Norse is a dead language, that was the father of modern languages like Icelandic, Swedish, Norwegian, Danish, Faroese and Elfdalian. Popularly known as the language that vikings spoke.
      </p>
    </section>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on \`"A Concise Dictionary of Old Icelandic\`" dictionary.
        </p>
        <p>
          Geir Zoëga attempted to to make an easier-to-approach version of the more full Cleasby - Vigfusson dictionary
        </p>
        <p>
          It was published in 1910, which leads to there being many public domain versions of the book available.
        </p>
      </section>
      <section
        className="section"
      >
        <h3>
          Old Norse language
        </h3>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/old-icelandic-zoega-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-icelandic-zoega"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-icelandic-zoega-abbreviations"
                rel="noopener noreferrer"
                target="_blank"
              >
                Abbreviations
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related packages
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/younger-futhark"
                rel="noopener noreferrer"
                target="_blank"
              >
                Younger Futhark Runes
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norse-alphabet"
                rel="noopener noreferrer"
                target="_blank"
              >
                Old Norse Alphabet
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norse-alphabet-sort"
                rel="noopener noreferrer"
                target="_blank"
              >
                Old Norse Alphabet Sort
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/gangleri"
                rel="noopener noreferrer"
                target="_blank"
              >
                Gangleri
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/teljari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Teljari
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/markari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Markari
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/spyrjari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Spyrjari
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                á
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                é
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                í
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                ó
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                ú
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                ý
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/th"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                þ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                æ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe3"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                œ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                ö
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2021
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
